src/System/TimedScopes/Correlation.cs

```diff
@@ -0,0 +1,302 @@
+		/// Key used to store if direct logging should be used
+		/// </summary>
+		private const string LoggingContextKey = "ULSLogging.directLogging";
+
+
+		/// <summary>
+		/// Get the current correlation if set
+		/// </summary>
+		/// <exception cref="InvalidCastException">Data exists on the CallContext
+		/// but it isn't a valid CorrelationData object.</exception>
+		public CorrelationData CurrentCorrelation
+		{
+			get
+			{
+				return GetCorrelationData();
+			}
+			private set
+			{
+				ICallContext existingContext = CallContext.ExistingCallContext();
+				if (existingContext != null)
+				{
+					if (value != null)
+					{
+						existingContext.Data[CorrelationContextKey] = value;
+					}
+					else
+					{
+						existingContext.Data.Remove(CorrelationContextKey);
+					}
+				}
+			}
+		}
+
+
+		/// <summary>
+		/// Should log directly to the underlying log handler
+		/// </summary>
+		public bool ShouldLogDirectly
+		{
+			get
+			{
+				CorrelationData correlation = CurrentCorrelation;
+				if (correlation != null && correlation.ShouldLogDirectly)
+				{
+					return true;
+				}
+
+				ICallContext existingContext = CallContext.ExistingCallContext();
+				if (existingContext != null)
+				{
+					object data;
+					return existingContext.Data.TryGetValue(LoggingContextKey, out data) && data != null;
+				}
+
+				return false;
+			}
+			set
+			{
+				CorrelationData correlation = CurrentCorrelation;
+				if (correlation != null)
+				{
+					correlation.ShouldLogDirectly = value;
+				}
+				else
+				{
+					ICallContext existingContext = CallContext.ExistingCallContext();
+					if (existingContext != null)
+					{
+						if (value)
+						{
+							existingContext.Data[LoggingContextKey] = new object();
+						}
+						else
+						{
+							existingContext.Data.Remove(LoggingContextKey);
+						}
+					}
+				}
+			}
+		}
+
+
+		/// <summary>
+		/// The correlation handler used to store the correlation
+		/// </summary>
+		/// <remarks>
+		/// Calling methods on this interface bypasses the
+		/// event handlers for correlation and commits the correlation
+		/// directly to the correlation handler.
+		/// </remarks>
+		public ICorrelationStorage CorrelationHandler { get; private set; }
+
+
+		/// <summary>
+		/// The correlation handler used to store the correlation
+		/// </summary>
+		/// <remarks>
+		/// Calling methods on this interface bypasses the
+		/// event handlers for correlation and commits the correlation
+		/// directly to the correlation handler.
+		/// </remarks>
+		public static ICallContextManager CallContextManager { get; private set; }
+
+
+		/// <summary>
+		/// The correlation handler used to store the correlation
+		/// </summary>
+		/// <remarks>
+		/// Calling methods on this interface bypasses the
+		/// event handlers for correlation and commits the correlation
+		/// directly to the correlation handler.
+		/// </remarks>
+		public static IMachineInformation MachineInformation { get; private set; }
```

Comment:
>private set; [](start = 62, length = 12)

Are the setters needed?

Reply:
They weren't, I removed them (for CallContextManager and StorageHandler as well).